File: sdk/javascript/src/offline.js
const { create, update, hash } = require('./block')

/**
 * Offline queue for creating FoodBlocks without network connectivity.
 * Blocks are stored locally and synced when connectivity is restored.
 *
 * See Section 5.5 of the whitepaper.
 */
class OfflineQueue {
  constructor() {
    this._blocks = []
  }

  /** Create a block and add it to the offline queue. */
  create(type, state = {}, refs = {}) {
    const block = create(type, state, refs)
    this._blocks.push(block)
    return block
  }

  /** Create an update block and add it to the offline queue. */
  update(previousHash, type, state = {}, refs = {}) {
    const block = update(previousHash, type, state, refs)
    this._blocks.push(block)
    return block
  }

  /** Get all queued blocks. */
  get blocks() {
    return [...this._blocks]
  }

  /** Number of queued blocks. */
  get length() {
    return this._blocks.length
  }

  /** Clear the queue (e.g. after successful sync). */
  clear() {
    this._blocks = []
  }

  /**
   * Sort blocks in dependency order for sync.
   * Blocks that reference other blocks in the queue are placed after their dependencies.
   */
  sorted() {
    const hashes = new Set(this._blocks.map(b => b.hash))
    const graph = new Map()

    for (const block of this._blocks) {
      const deps = []
      if (block.refs) {
        for (const ref of Object.values(block.refs)) {
          const refHashes = Array.isArray(ref) ? ref : [ref]
          for (const h of refHashes) {
            if (hashes.has(h)) deps.push(h)
          }
        }
      }
      graph.set(block.hash, deps)
    }

    // Topological sort
    const visited = new Set()
    const result = []

    const visit = (hash) => {
      if (visited.has(hash)) return
      visited.add(hash)
      for (const dep of (graph.get(hash) || [])) {
        visit(dep)
      }
      const block = this._blocks.find(b => b.hash === hash)
      if (block) result.push(block)
    }

    for (const block of this._blocks) {
      visit(block.hash)
    }

    return result
  }

  /**
   * Sync queued blocks to a remote server.
   *
   * @param {string} url - The FoodBlock server URL (e.g. 'http://localhost:3111')
   * @param {object} [opts] - { fetch } for custom fetch implementation
   * @returns {object} - { inserted: string[], skipped: string[], failed: { hash, error }[] }
   */
  async sync(url, opts = {}) {
    const fetchFn = opts.fetch || globalThis.fetch
    const sorted = this.sorted()

    const response = await fetchFn(`${url}/blocks/batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ blocks: sorted })
    })

    if (!response.ok) {
      throw new Error(`Sync failed: ${response.status} ${response.statusText}`)
    }

    const result = await response.json()
    this.clear()
    return result
  }
}

/**
 * Create a new offline queue.
 * @returns {OfflineQueue}
 */
function offlineQueue() {
  return new OfflineQueue()
}

module.exports = { offlineQueue, OfflineQueue }

File: sdk/go/offline.go
package foodblock

import "sort"

// OfflineQueue stores blocks created offline for later sync.
type OfflineQueue struct {
	blocks []Block
}

// NewOfflineQueue creates a new offline queue.
func NewOfflineQueue() *OfflineQueue {
	return &OfflineQueue{}
}

// Create creates a block and adds it to the offline queue.
func (q *OfflineQueue) Create(typ string, state, refs map[string]interface{}) Block {
	block := Create(typ, state, refs)
	q.blocks = append(q.blocks, block)
	return block
}

// Update creates an update block and adds it to the offline queue.
func (q *OfflineQueue) Update(previousHash, typ string, state, refs map[string]interface{}) Block {
	block := Update(previousHash, typ, state, refs)
	q.blocks = append(q.blocks, block)
	return block
}

// Blocks returns a copy of all queued blocks.
func (q *OfflineQueue) Blocks() []Block {
	result := make([]Block, len(q.blocks))
	copy(result, q.blocks)
	return result
}

// Len returns the number of queued blocks.
func (q *OfflineQueue) Len() int {
	return len(q.blocks)
}

// Clear empties the queue (e.g. after successful sync).
func (q *OfflineQueue) Clear() {
	q.blocks = nil
}

// Sorted returns blocks in dependency order for sync.
// Blocks that reference other blocks in the queue are placed after their dependencies.
func (q *OfflineQueue) Sorted() []Block {
	hashes := make(map[string]bool)
	for _, b := range q.blocks {
		hashes[b.Hash] = true
	}

	// Build dependency graph
	graph := make(map[string][]string)
	for _, block := range q.blocks {
		var deps []string
		for _, ref := range block.Refs {
			switch v := ref.(type) {
			case string:
				if hashes[v] {
					deps = append(deps, v)
				}
			case []interface{}:
				for _, item := range v {
					if s, ok := item.(string); ok && hashes[s] {
						deps = append(deps, s)
					}
				}
			}
		}
		graph[block.Hash] = deps
	}

	// Topological sort
	visited := make(map[string]bool)
	var result []Block
	blockMap := make(map[string]Block)
	for _, b := range q.blocks {
		blockMap[b.Hash] = b
	}

	var visit func(string)
	visit = func(hash string) {
		if visited[hash] {
			return
		}
		visited[hash] = true
		for _, dep := range graph[hash] {
			visit(dep)
		}
		if b, ok := blockMap[hash]; ok {
			result = append(result, b)
		}
	}

	// Sort hashes for deterministic ordering
	sortedHashes := make([]string, 0, len(q.blocks))
	for _, b := range q.blocks {
		sortedHashes = append(sortedHashes, b.Hash)
	}
	sort.Strings(sortedHashes)

	for _, hash := range sortedHashes {
		visit(hash)
	}

	return result
}
